language/move-prover/tests/sources/functional/is_signer.move

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public fun f6(account: &signer) { // This function produces no error because it satisfies f5's requires condition.
+        assert(Signer::address_of(account) == @0x7, 1);
+        f5();
+    }
+
+    public fun f7_incorrect() {
+        // This function does not satisfy f5's requires condition.
+        f5();
+    }
+
+
+    // ------------------------------
+    // Simple access control examples
+    // ------------------------------
+
+    spec fun hasPermissionAddr(addr: address): bool;
+    fun hasPermission(_account: &signer): bool { true }
+    spec hasPermission {
+        pragma opaque;
+        pragma verify=false;
+        ensures result == hasPermissionAddr(Signer::spec_address_of(_account));
+    }
+
+    public fun g_incorrect(_a: &signer, _b: &signer) {
+        spec {
+            assert exists addr:address: hasPermissionAddr(addr);
+        };
+        // privileged operation
+    }
+
+    public fun g1(a: &signer, _b: &signer) {
+        if(hasPermission(a))
+        {
+            spec {
+                assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+            };
+            // privileged operation
+        }
+    }
+
+    public fun g2(a: &signer, _b: &signer) {
+        assert (hasPermission(a), 1);
+        spec {
+            assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+        };
+        // privileged operation
+    }
+
+    public fun g3(a: &signer, _b: &signer) {
+        assert (hasPermission(a), 1);
+        helper()
+    }
+
+    public fun helper() {
+        spec {
+            assert exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+        };
+        // privileged operation
+    }
+    spec helper {
+        requires exists addr:address: (Signer::is_signer(addr) && hasPermissionAddr(addr));
+    }
+
+
+    // -----------------------------------
+    // Access control for Counter resource
+    // -----------------------------------
+
+    struct Counter has key { i: u64 }
+
+    public fun ADMIN_ADDRESS(): address {
+        @0x7
+    }
+
+    const AUTH_FAILED: u64 = 1;
+
+    public fun publish(account: &signer) {
+        spec { assume Signer::is_signer(Signer::spec_address_of(account)); };
+
+        assert(Signer::address_of(account) == ADMIN_ADDRESS(), AUTH_FAILED);
+        move_to(account, Counter { i: 0 });
+    }
+
+    public fun get_counter(): u64 acquires Counter {
+        borrow_global<Counter>(ADMIN_ADDRESS()).i
+    }
+
+    public fun increment(account: &signer) acquires Counter {
+        assert(Signer::address_of(account) == ADMIN_ADDRESS(), AUTH_FAILED); // permission check
+        let c_ref = &mut borrow_global_mut<Counter>(ADMIN_ADDRESS()).i;
+        *c_ref = *c_ref + 1;
+    }
+
+    // This function is incorrect because it omits the permission check.
+    public fun increment_incorrect(_account: &signer) acquires Counter {
```

Comment:
This function violates the invariant below.

Reply:
So from this example, is_signer is mainly there to connect a global invariant -- where we can't talk about signers normally but just addresses -- to the code which has signers. Would it also be used somewhere else than invariants?

@DavidLDill